File: src/Billing/CampaignService.php
<?php
declare(strict_types=1);
namespace App\Billing;
use App\Infrastructure\{Database,Outbox};
final class CampaignService
{
    public function __construct(private Database $db, private Outbox $outbox, private Wallet $wallet) {}
    /** Create a campaign. */
    public function create(array $input, string $actor): array
    {
        $name = trim((string)($input['name'] ?? ''));
        $param = trim((string)($input['start_parameter'] ?? ''));
        if ($name === '' || mb_strlen($name) > 255) throw new BillingError('Название: 1–255 символов.');
        if (!preg_match('/^[a-zA-Z0-9_-]{2,64}$/D', $param)) throw new BillingError('Параметр: 2–64 символа a-z, 0-9, _ или -.');
        $type = (string)($input['bonus_type'] ?? 'balance');
        if (!in_array($type, ['balance', 'subscription', 'tariff'], true)) throw new BillingError('Некорректный тип бонуса.');
        $id = Database::id();
        $this->db->execute(
            'INSERT INTO advertising_campaigns(id,name,start_parameter,bonus_type,balance_bonus_kopeks,subscription_duration_days,subscription_traffic_gb,subscription_device_limit,plan_id,is_active,partner_user_id,created_by,created_at) VALUES(?,?,?,?,?,?,?,?,?,1,?,?,?)',
            [$id, $name, $param, $type, (int)($input['balance_bonus_kopeks'] ?? 0), $input['subscription_duration_days'] !== '' ? (int)$input['subscription_duration_days'] : null, $input['subscription_traffic_gb'] !== '' ? (int)$input['subscription_traffic_gb'] : null, $input['subscription_device_limit'] !== '' ? (int)$input['subscription_device_limit'] : null, $input['plan_id'] !== '' ? $input['plan_id'] : null, $input['partner_user_id'] !== '' ? $input['partner_user_id'] : null, $actor, time()]
        );
        $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'campaign.created', $id, time()]);
        return $this->db->one('SELECT * FROM advertising_campaigns WHERE id=?', [$id]);
    }
    public function list(): array
    {
        return $this->db->all('SELECT * FROM advertising_campaigns ORDER BY created_at DESC');
    }
    /** Register a user via campaign start parameter. Grants the bonus once. */
    public function register(string $userId, string $startParameter): ?array
    {
        $campaign = $this->db->one('SELECT * FROM advertising_campaigns WHERE start_parameter=? AND is_active=1', [$startParameter]);
        if (!$campaign) return null;
        return $this->db->transaction(function () use ($userId, $campaign) {
            if ($this->db->one('SELECT id FROM advertising_campaign_registrations WHERE campaign_id=? AND user_id=?', [$campaign['id'], $userId])) return $campaign;
            $this->db->execute('INSERT INTO advertising_campaign_registrations(id,campaign_id,user_id,bonus_granted,created_at) VALUES(?,?,?,0,?)', [Database::id(), $campaign['id'], $userId, time()]);
            $this->grant($userId, $campaign);
            $this->db->execute('UPDATE advertising_campaign_registrations SET bonus_granted=1 WHERE campaign_id=? AND user_id=?', [$campaign['id'], $userId]);
            return $campaign;
        });
    }
    private function grant(string $userId, array $campaign): void
    {
        $type = $campaign['bonus_type'];
        if ($type === 'balance' && (int)$campaign['balance_bonus_kopeks'] > 0) {
            $this->wallet->credit($userId, (int)$campaign['balance_bonus_kopeks'], 'promo_credit', 'Бонус кампании: '.$campaign['name']);
        } elseif ($type === 'subscription' && (int)$campaign['subscription_duration_days'] > 0) {
            $days = (int)$campaign['subscription_duration_days'];
            $sub = $this->db->one("SELECT * FROM subscriptions WHERE user_id=? AND status IN ('active','trial') ORDER BY created_at DESC LIMIT 1", [$userId]);
            if ($sub) {
                $base = max(time(), (int)$sub['expires_at']);
                $this->db->execute("UPDATE subscriptions SET expires_at=?,status='active' WHERE id=?", [$base + $days * 86400, $sub['id']]);
                $this->outbox->enqueue('subscription.extend', 'campaign-extend:'.$campaign['id'].':'.$sub['id'], ['subscription_id' => $sub['id']]);
            } else {
                $plan = $this->db->one('SELECT * FROM plans WHERE id=?', [$campaign['plan_id']]);
                $subId = Database::id();
                $now = time();
                $this->db->execute(
                    "INSERT INTO subscriptions(id,order_id,user_id,status,expires_at,created_at,plan_id,traffic_limit_gb,device_limit,is_trial,start_date) VALUES(?,NULL,?,'active',?,?,?,?,?,0,?)",
                    [$subId, $userId, $now + $days * 86400, $now, $plan['id'] ?? null, $plan ? (int)$plan['traffic_bytes'] / 1073741824 : 0, $plan ? (int)$plan['devices'] : 1, $now]
                );
                $this->outbox->enqueue('subscription.provision', 'provision:'.$subId, ['subscription_id' => $subId]);
            }
        }
    }
}

File: rails_backend/app/services/billing/campaign_service.rb
module Billing
  class CampaignService
    def create(input:, actor:)
      attrs = input.to_h.symbolize_keys
      name = attrs[:name].to_s.strip
      parameter = attrs[:start_parameter].to_s.strip
      raise BillingError, "Название: 1–255 символов." unless name.length.between?(1, 255)
      raise BillingError, "Параметр: 2–64 символа a-z, 0-9, _ или -." unless parameter.match?(/\A[a-zA-Z0-9_-]{2,64}\z/)
      type = attrs[:bonus_type].presence || "balance"
      raise BillingError, "Некорректный тип бонуса." unless %w[balance subscription].include?(type)
      balance = attrs[:balance_bonus_kopeks].to_i
      days = attrs[:subscription_duration_days].to_i
      plan_id = attrs[:plan_id].to_s
      if type == "balance"
        raise BillingError, "Бонус: 1–1 000 000 ₽." unless balance.between?(1, 100_000_000)
      else
        raise BillingError, "Для подписки укажите 1–3650 дней и активный тариф." unless days.between?(1, 3650) && Plan.active.exists?(id: plan_id)
      end
      now = Time.now.to_i
      AdvertisingCampaign.transaction do
        campaign = AdvertisingCampaign.create!(
          id: Infrastructure::IdGenerator.call, name: name, start_parameter: parameter,
          bonus_type: type, balance_bonus_kopeks: type == "balance" ? balance : 0,
          subscription_duration_days: type == "subscription" ? days : nil,
          subscription_traffic_gb: nil, subscription_device_limit: nil,
          plan_id: type == "subscription" ? plan_id : nil, is_active: 1,
          partner_user_id: attrs[:partner_user_id].presence, created_by: actor, created_at: now
        )
        audit(actor, "campaign.created", campaign.id, now)
        campaign
      end
    rescue ActiveRecord::RecordNotUnique
      raise BillingError, "Параметр кампании уже используется."
    end

    def list
      AdvertisingCampaign.order(created_at: :desc).limit(500)
    end

    def register(user_id, start_parameter)
      AdvertisingCampaign.transaction do
        campaign = AdvertisingCampaign.lock.find_by(start_parameter: start_parameter.to_s, is_active: 1)
        next nil unless campaign
        user = User.find_by(id: user_id)
        raise BillingError, "Аккаунт недоступен." unless user && user.disabled.to_i.zero?
        existing = AdvertisingCampaignRegistration.find_by(campaign_id: campaign.id, user_id: user_id)
        next campaign if existing
        registration = AdvertisingCampaignRegistration.create!(id: Infrastructure::IdGenerator.call,
          campaign_id: campaign.id, user_id: user_id, bonus_granted: 0, created_at: Time.now.to_i)
        grant(user_id, campaign)
        registration.update!(bonus_granted: 1)
        campaign
      end
    end

    private

    def grant(user_id, campaign)
      if campaign.bonus_type == "balance" && campaign.balance_bonus_kopeks.to_i.positive?
        WalletService.new.credit(user_id, campaign.balance_bonus_kopeks.to_i, "promo_credit", "Бонус кампании: #{campaign.name}")
      elsif campaign.bonus_type == "subscription" && campaign.subscription_duration_days.to_i.positive?
        days = campaign.subscription_duration_days.to_i
        subscription = Subscription.where(user_id: user_id, status: %w[active trial]).order(created_at: :desc).lock.first
        if subscription
          expiry = [Time.now.to_i, subscription.expires_at.to_i].max + days * 86_400
          subscription.update!(expires_at: expiry, status: "active", lifecycle_status: "active", updated_at: Time.now.to_i)
          Infrastructure::OutboxService.new.enqueue("subscription.extend", "campaign-extend:#{campaign.id}:#{subscription.id}", { subscription_id: subscription.id })
        else
          plan = Plan.find_by(id: campaign.plan_id)
          raise BillingError, "Тариф кампании больше недоступен." unless plan
          now = Time.now.to_i
          sub = Subscription.create!(id: Infrastructure::IdGenerator.call, order_id: nil, user_id: user_id,
            status: "provisioning", lifecycle_status: "pending", expires_at: now + days * 86_400,
            created_at: now, updated_at: now, starts_at: now, plan_id: plan.id,
            traffic_limit_gb: plan.traffic_bytes.to_i / 1.gigabyte,
            traffic_limit_bytes: plan.traffic_bytes.to_i, device_limit: plan.devices,
            is_trial: 0, start_date: now)
          Infrastructure::OutboxService.new.enqueue("subscription.provision", "provision:#{sub.id}", { subscription_id: sub.id })
        end
      else
        raise BillingError, "Бонус кампании не настроен."
      end
    end

    def audit(actor, action, subject, now)
      AuditLog.create!(id: Infrastructure::IdGenerator.call, actor: actor, action: action, subject: subject, created_at: now)
    end
  end
end
